### PR TITLE
Multilingual support

### PR DESCRIPTION
**Pull request by madaian**
Based on #311 
- course can have multiple languages
- can store titles and URLs in multiple languages in `|en:Chapter|fi:Luku|` format
- renders text fields in currently selected language
- course menu has buttons to change the language

URLs are not validated, as they are handled as regular text

### CODE REVIEW

**Comment by Plustest** [comment]
Can one of the admins verify this patch?

File: course/models.py [hunk 1]
```diff
@@ -356,7 +356,15 @@ def clean(self):
...
             errors['archive_time'] = _("Archive time must be later than lifesupport time.")
-        if not self.is_valid_language(self.language):
+        if self.language.startswith("|"):
+            langs = list(filter(None, self.language.split("|"))) # remove pipes & empty strings
+            for lang in langs:
+                if not self.is_valid_language(lang):
+                    if "language" in errors:
+                        errors['language'] = "%s%s" % (errors['language'], (", " + lang))
+                    else:
+                        errors['language'] = _("Language code(s) missing from settings: " + lang)
```
Comment (raphendyr) [inline course/models.py]: This doesn't go well in translation. Text inside function `_()` is translated, but it's now different based on language. `_("plaa plaa: ") + lang` should be good enough here.

File: course/templatetags/course.py [hunk 2]
```diff
@@ -38,6 +39,32 @@ def group_select(context):
...
+def parse_localization(entry):
+    """
+    Picks the currently selected language's value from
+    |lang:value|lang:value| -format text.
+    """
+    text = ""
+    if isinstance(entry, str):
+        text = entry
+    else:
+        text = str(entry) # if not a string, assume entry is a model
```
Comment (raphendyr) [inline course/templatetags/course.py]: `text  = entry if isinstance(entry, str) else str(entry)`. Though, `str(a_string_object)` would work too.

Other solution would be setting entry if it's not str:
`if not isinstance(entry, str): entry = str(entry)`

File: course/viewbase.py [hunk 3]
```diff
@@ -66,7 +67,13 @@ def get_resource_objects(self):
 
             # Apply course instance language.
             if self.instance.language:
-                translation.activate(self.instance.language)
+                lang = self.instance.language
+                if lang.startswith("|"):
+                    active = get_language()
+                    if "|" + active + "|" in lang:
+                        translation.activate(active)
```
Comment (raphendyr) [inline course/viewbase.py]: What happens if active language is not in the list?

File: edit_course/operations/configure.py [hunk 4]
```diff
@@ -239,8 +251,18 @@ def configure_content(instance, url):
...
+    if "lang" in config:
+        langs = config["lang"]
+        if isinstance(langs, list):
+            formatted = ""
+            for lang in langs:
+                if instance.is_valid_language(lang):
+                    formatted += ("|" + lang)
+            if len(formatted) > 1:
+                formatted += "|"
+                instance.language = str(formatted)
```
Comment (raphendyr) [inline edit_course/operations/configure.py]: Would ```langs = [lang for lang in langs if instance.is_valid_language(lang)]
if langs:
	instance.language = "|{}|".format("|".join(langs))``` be a bit cleaner?

File: edit_course/operations/configure.py [hunk 5]
```diff
@@ -68,6 +68,18 @@ def parse_float(value, errors):
...
+def format_localization(element):
+    """
+    Parse localised elements into |lang:val|lang:val| -format strings
+    """
+    if isinstance(element, dict):
+        concat = ""
+        for pair in element.items():
+            concat += ("|" + ":".join(pair))
+        concat += "|"
+        return concat
```
Comment (raphendyr) [inline edit_course/operations/configure.py]: maybe?

```
strings = ("{}:{}".format(k, v) for k, v in element,items())
return "|{}|".format("|".join(strings))
```

File: edit_course/operations/configure.py [hunk 6]
```diff
@@ -276,7 +298,7 @@ def configure_content(instance, url):
         if not "name" in c:
             errors.append(_("Category requires a name."))
             continue
-        category, flag = instance.categories.get_or_create(name=str(c["name"]))
+        category, flag = instance.categories.get_or_create(name=format_localization(c["name"]))
```
Comment (raphendyr) [inline edit_course/operations/configure.py]: This requires more looking into. Technically categories should have `key/slug` and `display string` instead of them both being the same field. We need to look into this a bit more to see if this works as expected.

File: exercise/templates/exercise/exercise_base.html [hunk 7]
```diff
@@ -4,7 +4,7 @@
 {% load exercise %}
 {% load apps %}
 
-{% block title %}{{ current.name }} | {{ block.super }}{% endblock %}
+{% block title %}{{ current.name|parse_localization }} | {{ block.super }}{% endblock %}
```
Comment (raphendyr) [inline exercise/templates/exercise/exercise_base.html]: Need to use translation block with variable and run the filter there

File: course/viewbase.py (hunk 3 above)
Comment (madaian) [reply course/viewbase.py]: Fixed to fall back to the first language discovered in the course's settings so that the multilingual fields won't just disappear.

File: exercise/templates/exercise/exercise_base.html (hunk 7 above)
Comment (madaian) [reply exercise/templates/exercise/exercise_base.html]: I don't understand what the problem here is, it seems to be working fine?

Comment (raphendyr) [reply exercise/templates/exercise/exercise_base.html]: I see. Django works interestingly on that matter, so it really does work. Django strores everything inside `{{}}` as the "key" (think as variable) and uses that with `%` python format to expand the variables in translation string. Only downside from current is that the translation contains `current.name|parse_localization` instead of `name` for example. Thus in real life, it could be obscuring detail for the translator.

Relevant parts in django source:
 * store vars: https://github.com/django/django/blob/master/django/templatetags/i18n.py#L119
 * string format: https://github.com/django/django/blob/master/django/templatetags/i18n.py#L164

So in short, current code is fine and ok.
